### ISSUE TITLE
Data race concurrent map R/W

### ISSUE THREAD

**Issue by etlfg** [mattermost/mattermost-server#13797]
#### Summary
Data race concurrent map R/W

#### Steps to reproduce
I have trouble reproducing It. The server is in version 5.19.0 and have shutdown randomly several times already.

#### Expected behavior
The server should run without untimely crashing. 

#### Observed behavior (that appears unintentional) 
We have hundreds of users on a unique server running in docker behind a reverse proxy. Is has been running for months without issue in version 4.8. We updated It to Its version 5.19 ( (mattermost/mattermost-prod-app:5.19.0) and since, experienced 3-4 crashes at different time of the day.
There's a thousands of lines trace in the logs happening in less than a second. But It's hard to read and to pick relevant information from It. Below is an extract from the first line. 

```
2020-01-30T09:45:16.324678286Z fatal error: concurrent map read and map write
2020-01-30T09:45:16.327773959Z 
2020-01-30T09:45:16.327797582Z goroutine 2820 [running]:
2020-01-30T09:45:16.327805359Z runtime.throw(0x190d5b1, 0x21)
2020-01-30T09:45:16.327811731Z 	runtime/panic.go:774 +0x72 fp=0xc0029c3b20 sp=0xc0029c3af0 pc=0x431742
2020-01-30T09:45:16.327818555Z runtime.mapaccess1_faststr(0x16bb700, 0xc009a4da70, 0x18e0304, 0x11, 0xc005f9e0b8)
2020-01-30T09:45:16.327824763Z 	runtime/map_faststr.go:21 +0x44f fp=0xc0029c3b90 sp=0xc0029c3b20 pc=0x414a4f
2020-01-30T09:45:16.327831009Z github.com/mattermost/mattermost-server/v5/app.(*App).buildFullPushNotificationMessage(0xc0001ec380, 0xc00229b640, 0x7, 0xc0074e9200, 0xc005eb58c0, 0xc004d0a4e0, 0xc008df6110, 0xe, 0xc0060592f0, 0xa, ...)
2020-01-30T09:45:16.327838639Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:507 +0x320 fp=0xc0029c3c50 sp=0xc0029c3b90 pc=0x105a8d0
2020-01-30T09:45:16.327844375Z github.com/mattermost/mattermost-server/v5/app.(*App).BuildPushNotificationMessage(0xc0001ec380, 0xc00229b640, 0x7, 0xc0074e9200, 0xc005eb58c0, 0xc004d0a4e0, 0xc008df6110, 0xe, 0xc0060592f0, 0xa, ...)
2020-01-30T09:45:16.327851117Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:453 +0x14b fp=0xc0029c3cd8 sp=0xc0029c3c50 pc=0x105a1db
2020-01-30T09:45:16.327860117Z github.com/mattermost/mattermost-server/v5/app.(*App).sendPushNotificationSync(0xc0001ec380, 0xc0074e9200, 0xc005eb58c0, 0xc004d0a4e0, 0xc008df6110, 0xe, 0xc0060592f0, 0xa, 0x0, 0x0, ...)
2020-01-30T09:45:16.327867338Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:57 +0x102 fp=0xc0029c3d60 sp=0xc0029c3cd8 pc=0x1054d52
2020-01-30T09:45:16.327873832Z github.com/mattermost/mattermost-server/v5/app.(*App).pushNotificationWorker(0xc0001ec380, 0xc003306f00)
2020-01-30T09:45:16.327880024Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:261 +0x514 fp=0xc0029c3f88 sp=0xc0029c3d60 pc=0x10578b4
2020-01-30T09:45:16.327893889Z github.com/mattermost/mattermost-server/v5/app.(*App).StartPushNotificationsHubWorkers.func1()
2020-01-30T09:45:16.327900633Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:286 +0x33 fp=0xc0029c3fa8 sp=0xc0029c3f88 pc=0x10f09e3
2020-01-30T09:45:16.327906904Z github.com/mattermost/mattermost-server/v5/app.(*Server).Go.func1(0xc002c61360, 0xc002f6aa80)
2020-01-30T09:45:16.327912943Z 	github.com/mattermost/mattermost-server/v5@/app/server.go:394 +0x27 fp=0xc0029c3fd0 sp=0xc0029c3fa8 pc=0x10f5a27
2020-01-30T09:45:16.327919259Z runtime.goexit()
2020-01-30T09:45:16.327925121Z 	runtime/asm_amd64.s:1357 +0x1 fp=0xc0029c3fd8 sp=0xc0029c3fd0 pc=0x4613e1
2020-01-30T09:45:16.327933720Z created by github.com/mattermost/mattermost-server/v5/app.(*Server).Go
2020-01-30T09:45:16.327940264Z 	github.com/mattermost/mattermost-server/v5@/app/server.go:393 +0x53
2020-01-30T09:45:16.327946402Z 
2020-01-30T09:45:16.327952028Z goroutine 1 [chan receive, 35 minutes]:
2020-01-30T09:45:16.327958059Z github.com/mattermost/mattermost-server/v5/cmd/mattermost/commands.runServer(0x1be5500, 0xc002845000, 0xc0001b0000, 0xc0001b8180, 0x0, 0x0)
2020-01-30T09:45:16.327964451Z 	github.com/mattermost/mattermost-server/v5@/cmd/mattermost/commands/server.go:95 +0x350
2020-01-30T09:45:16.327970638Z github.com/mattermost/mattermost-server/v5/cmd/mattermost/commands.serverCmdF(0x2876540, 0x28cade0, 0x0, 0x0, 0x0, 0x0)
2020-01-30T09:45:16.327996604Z 	github.com/mattermost/mattermost-server/v5@/cmd/mattermost/commands/server.go:53 +0x1af
2020-01-30T09:45:16.328003937Z github.com/spf13/cobra.(*Command).execute(0x2876540, 0xc0000b4190, 0x0, 0x0, 0x2876540, 0xc0000b4190)
2020-01-30T09:45:16.328010470Z 	github.com/spf13/cobra@v0.0.5/command.go:826 +0x460
2020-01-30T09:45:16.328016468Z github.com/spf13/cobra.(*Command).ExecuteC(0x2876540, 0x0, 0x9a380, 0xc0000aa058)
2020-01-30T09:45:16.328022521Z 	github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
2020-01-30T09:45:16.328028407Z github.com/spf13/cobra.(*Command).Execute(...)
2020-01-30T09:45:16.328034188Z 	github.com/spf13/cobra@v0.0.5/command.go:864
2020-01-30T09:45:16.328040145Z github.com/mattermost/mattermost-server/v5/cmd/mattermost/commands.Run(...)
2020-01-30T09:45:16.328047125Z 	github.com/mattermost/mattermost-server/v5@/cmd/mattermost/commands/root.go:15
2020-01-30T09:45:16.328052344Z main.main()
2020-01-30T09:45:16.328058201Z 	github.com/mattermost/mattermost-server/v5@/cmd/mattermost/main.go:30 +0x87
2020-01-30T09:45:16.328067189Z 
2020-01-30T09:45:16.328073322Z goroutine 5 [syscall, 35 minutes]:
2020-01-30T09:45:16.328079190Z os/signal.signal_recv(0x0)
2020-01-30T09:45:16.328085935Z 	runtime/sigqueue.go:147 +0x9c
2020-01-30T09:45:16.328090879Z os/signal.loop()
2020-01-30T09:45:16.328096681Z 	os/signal/signal_unix.go:23 +0x22
2020-01-30T09:45:16.328103628Z created by os/signal.init.0
2020-01-30T09:45:16.328112865Z 	os/signal/signal_unix.go:29 +0x41
2020-01-30T09:45:16.328119146Z 
2020-01-30T09:45:16.328125011Z goroutine 2790 [chan receive, 35 minutes]:
2020-01-30T09:45:16.328130981Z github.com/mattermost/mattermost-server/v5/app.(*App).pushNotificationWorker(0xc0001ec380, 0xc0033063c0)
2020-01-30T09:45:16.328137234Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:254 +0x72
2020-01-30T09:45:16.328143402Z github.com/mattermost/mattermost-server/v5/app.(*App).StartPushNotificationsHubWorkers.func1()
2020-01-30T09:45:16.328149616Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:286 +0x33
2020-01-30T09:45:16.328155558Z github.com/mattermost/mattermost-server/v5/app.(*Server).Go.func1(0xc002c60fa0, 0xc002f6aa80)
2020-01-30T09:45:16.328161551Z 	github.com/mattermost/mattermost-server/v5@/app/server.go:394 +0x27
2020-01-30T09:45:16.328167965Z created by github.com/mattermost/mattermost-server/v5/app.(*Server).Go
2020-01-30T09:45:16.328174155Z 	github.com/mattermost/mattermost-server/v5@/app/server.go:393 +0x53
2020-01-30T09:45:16.328180049Z 
2020-01-30T09:45:16.328185585Z goroutine 2789 [chan receive, 35 minutes]:
2020-01-30T09:45:16.328194933Z github.com/mattermost/mattermost-server/v5/app.(*App).pushNotificationWorker(0xc0001ec380, 0xc003306360)
2020-01-30T09:45:16.328197524Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:254 +0x72
2020-01-30T09:45:16.328204962Z github.com/mattermost/mattermost-server/v5/app.(*App).StartPushNotificationsHubWorkers.func1()
2020-01-30T09:45:16.328209598Z 	github.com/mattermost/mattermost-server/v5@/app/notification_push.go:286 +0x33
2020-01-30T09:45:16.328215759Z github.com/mattermost/mattermost-server/v5/app.(*Server).Go.func1(0xc002c60f80, 0xc002f6aa80)
2020-01-30T09:45:16.328221803Z 	github.com/mattermost/mattermost-server/v5@/app/server.go:394 +0x27
2020-01-30T09:45:16.328230524Z created by github.com/mattermost/mattermost-server/v5/app.(*Server).Go
2020-01-30T09:45:16.328233907Z 	github.com/mattermost/mattermost-server/v5@/app/server.go:393 +0x53
2020-01-30T09:45:16.328239867Z 
2020-01-30T09:45:16.328245510Z goroutine 8527 [IO wait]:
2020-01-30T09:45:16.328253738Z internal/poll.runtime_pollWait(0x7fcfc7bb5888, 0x72, 0xffffffffffffffff)
2020-01-30T09:45:16.328260176Z 	runtime/netpoll.go:184 +0x55
2020-01-30T09:45:16.328266103Z internal/poll.(*pollDesc).wait(0xc007ecf298, 0x72, 0x2000, 0x2000, 0xffffffffffffffff)
2020-01-30T09:45:16.328272130Z 	internal/poll/fd_poll_runtime.go:87 +0x45
2020-01-30T09:45:16.328278136Z internal/poll.(*pollDesc).waitRead(...)
2020-01-30T09:45:16.328283991Z 	internal/poll/fd_poll_runtime.go:92
2020-01-30T09:45:16.328289827Z internal/poll.(*FD).Read(0xc007ecf280, 0xc00833a000, 0x2000, 0x2000, 0x0, 0x0, 0x0)
2020-01-30T09:45:16.328299583Z 	internal/poll/fd_unix.go:169 +0x1cf
2020-01-30T09:45:16.328305865Z net.(*netFD).Read(0xc007ecf280, 0xc00833a000, 0x2000, 0x2000, 0x10, 0x10, 0x1704120)
2020-01-30T09:45:16.328312722Z 	net/fd_unix.go:202 +0x4f
2020-01-30T09:45:16.328317899Z net.(*conn).Read(0xc003f2c770, 0xc00833a000, 0x2000, 0x2000, 0x0, 0x0, 0x0)
2020-01-30T09:45:16.328323944Z 	net/net.go:184 +0x68
2020-01-30T09:45:16.328329840Z bufio.(*Reader).fill(0xc0082c6540)
2020-01-30T09:45:16.328335762Z 	bufio/bufio.go:100 +0x103
2020-01-30T09:45:16.328341668Z bufio.(*Reader).Peek(0xc0082c6540, 0x2, 0x10fba5d, 0xc002f6aa80, 0xc008376990, 0x20271097da3, 0x28ad9c0)
2020-01-30T09:45:16.328347770Z 	bufio/bufio.go:138 +0x4f
2020-01-30T09:45:16.328353789Z github.com/gorilla/websocket.(*Conn).read(0xc00833c000, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
2020-01-30T09:45:16.328359754Z 	github.com/gorilla/websocket@v1.4.1/conn.go:370 +0x40
2020-01-30T09:45:16.328365863Z github.com/gorilla/websocket.(*Conn).advanceFrame(0xc00833c000, 0xa, 0x0, 0x0)
2020-01-30T09:45:16.328371802Z 	github.com/gorilla/websocket@v1.4.1/conn.go:798 +0x5c
2020-01-30T09:45:16.328378466Z github.com/gorilla/websocket.(*Conn).NextReader(0xc00833c000, 0xc007ecf280, 0xbf84c555dc82be83, 0x197e59425c, 0xc007840e60, 0x40e918)
2020-01-30T09:45:16.328384139Z 	github.com/gorilla/websocket@v1.4.1/conn.go:980 +0xa0
2020-01-30T09:45:16.328390219Z github.com/gorilla/websocket.(*Conn).ReadJSON(0xc00833c000, 0x1691200, 0xc0082f2410, 0x28ad9c0, 0x0)
2020-01-30T09:45:16.328396383Z 	github.com/gorilla/websocket@v1.4.1/json.go:50 +0x2f
2020-01-30T09:45:16.328402468Z github.com/mattermost/mattermost-server/v5/app.(*WebConn).readPump(0xc0077778c0)
2020-01-30T09:45:16.328408477Z 	github.com/mattermost/mattermost-server/v5@/app/web_conn.go:144 +0x19f
2020-01-30T09:45:16.328414584Z github.com/mattermost/mattermost-server/v5/app.(*WebConn).Pump(0xc0077778c0)
2020-01-30T09:45:16.328420562Z 	github.com/mattermost/mattermost-server/v5@/app/web_conn.go:117 +0x79
2020-01-30T09:45:16.328426629Z github.com/mattermost/mattermost-server/v5/api4.connectWebSocket(0xc008015860, 0x1bc7160, 0xc0039e7730, 0xc007f7b000)
2020-01-30T09:45:16.328432762Z 	github.com/mattermost/mattermost-server/v5@/api4/websocket.go:39 +0x319
2020-01-30T09:45:16.328440101Z github.com/mattermost/mattermost-server/v5/web.Handler.ServeHTTP(0xc001714c90, 0x19bf070, 0x25326f8, 0x10, 0x100, 0x0, 0x0, 0x1bc7160, 0xc0039e7730, 0xc007f7b000)
2020-01-30T09:45:16.328446946Z 	github.com/mattermost/mattermost-server/v5@/web/handlers.go:163 +0x1b9e
2020-01-30T09:45:16.328453047Z github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1(0x1bc6be0, 0xc0072df6c0, 0xc007f7b000)
2020-01-30T09:45:16.328459416Z 	github.com/NYTimes/gziphandler@v1.1.1/gzip.go:336 +0x23f
2020-01-30T09:45:16.328465447Z net/http.HandlerFunc.ServeHTTP(0xc003a30540, 0x1bc6be0, 0xc0072df6c0, 0xc007f7b000)
2020-01-30T09:45:16.328471403Z 	net/http/server.go:2007 +0x44
2020-01-30T09:45:16.328480336Z github.com/gorilla/mux.(*Router).ServeHTTP(0xc002dfe180, 0x1bc6be0, 0xc0072df6c0, 0xc007f7ae00)
2020-01-30T09:45:16.328486736Z 	github.com/gorilla/mux@v1.7.3/mux.go:212 +0xe2
2020-01-30T09:45:16.328492600Z github.com/mattermost/mattermost-server/v5/app.(*RateLimiter).RateLimitHandler.func1(0x1bc6be0, 0xc0072df6c0, 0xc007f7ae00)
2020-01-30T09:45:16.328498826Z 	github.com/mattermost/mattermost-server/v5@/app/ratelimit.go:108 +0xc8
2020-01-30T09:45:16.328504823Z net/http.HandlerFunc.ServeHTTP(0xc0046f3580, 0x1bc6be0, 0xc0072df6c0, 0xc007f7ae00)
2020-01-30T09:45:16.328510696Z 	net/http/server.go:2007 +0x44
2020-01-30T09:45:16.328516511Z net/http.serverHandler.ServeHTTP(0xc004644460, 0x1bc6be0, 0xc0072df6c0, 0xc007f7ae00)
2020-01-30T09:45:16.328522440Z 	net/http/server.go:2802 +0xa4
2020-01-30T09:45:16.328530166Z net/http.(*conn).serve(0xc007f6a640, 0x1bcc620, 0xc008038100)
2020-01-30T09:45:16.328535942Z 	net/http/server.go:1890 +0x875
2020-01-30T09:45:16.328540107Z created by net/http.(*Server).Serve
2020-01-30T09:45:16.328545786Z 	net/http/server.go:2927 +0x38e
```

There's something else but I don't know If it's related. But since the update, we have observed new entries in the log about Notifications. The logs below show how It looks like :
```
2020-01-30T09:45:16.324657179Z {"level":"warn","ts":1580377516.3243523,"caller":"app/notification.go:293","msg":"Notification not sent","logSource":"notifications","ackId":"","type":"message","userId":"<obfuscated>","postId":"<obfuscated>","status":"Not Sent due to preferences"}
```

#### Possible fixes
It seems we didn't enable a given option in the configuration which causes these logs to happen. Does It matter(most) ?

It may be related to one of the upgrade notes there https://docs.mattermost.com/administration/important-upgrade-notes.html. For example, the one about notification on secure/insecure servers from version 5.16. But It shouldn't cause crashes, hu ?

#### Title
I can change the title to make It more explicit. But I don't really know what's It's about atm.

**Comment by agnivade** [comment]
Hi @etlfg, thank you for this report ! We are already tracking this with https://mattermost.atlassian.net/browse/MM-21378. 

In the mean while, there is a workaround, but I'm afraid it might not be very feasible. It is to disable push notifications with `SendPushNotifications: false`. 

The log line `"Notification not sent"` is related but not something to do with this issue. It gets generated if a message is not meant to be sent for push notification.

**Comment by elfshed** [comment]
> #### Summary
> 
> Data race concurrent map R/W

Two of our servers experienced the crash last night with 3000 lines written to the systemd journal in the timespan of 3 seconds. 

Version 5.19.1

**Comment by agnivade** [comment]
Hi @elfshed - we already have a PR up for this https://github.com/mattermost/mattermost-server/pull/13884. It's set for the 5.21 release. Alternatively, you can also do a custom build once it gets in to master.

**Comment by amyblais** [comment]
This is fixed for upcoming v5.22 release, scheduled for April 16th,
